gateway/ln-gateway/src/lightning/ldk.rs

```diff
@@ -0,0 +1,501 @@
+    CloseChannelsWithPeerResponse, CreateInvoiceRequest, CreateInvoiceResponse, EmptyResponse,
+    GetFundingAddressResponse, GetNodeInfoResponse, GetRouteHintsResponse, InterceptHtlcRequest,
+    InterceptHtlcResponse, PayInvoiceRequest, PayInvoiceResponse,
+};
+
+pub struct GatewayLdkClient {
+    /// The underlying lightning node.
+    node: Arc<ldk_node::Node>,
+
+    /// The client for querying data about the blockchain.
+    esplora_client: esplora_client::AsyncClient,
+
+    /// A handle to the task that seeds the HTLC stream.
+    /// TODO: This should be a shutdown sender instead, and we can discard the
+    /// handle.
+    htlc_stream_seeder_task_handle: tokio::task::JoinHandle<()>,
+
+    /// The HTLC stream, until it is taken by calling
+    /// `ILnRpcClient::route_htlcs`.
+    htlc_stream_receiver_or:
+        Option<tokio::sync::mpsc::Receiver<Result<InterceptHtlcRequest, Status>>>,
+}
+
+impl std::fmt::Debug for GatewayLdkClient {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("GatewayLdkClient").finish_non_exhaustive()
+    }
+}
+
+impl GatewayLdkClient {
+    /// Creates a new `GatewayLdkClient` instance and starts the underlying
+    /// lightning node. All resources, including the lightning node, will be
+    /// cleaned up when the returned `GatewayLdkClient` instance is dropped.
+    /// There's no need to manually stop the node.
+    pub fn new(
+        data_dir: &Path,
+        esplora_server_url: &str,
+        network: Network,
+        lightning_port: u16,
+    ) -> anyhow::Result<Self> {
+        let mut node_builder = ldk_node::Builder::from_config(ldk_node::Config {
+            network,
+            listening_addresses: Some(vec![SocketAddress::TcpIpV4 {
+                addr: [0, 0, 0, 0],
+                port: lightning_port,
+            }]),
+            onchain_wallet_sync_interval_secs: 10,
+            wallet_sync_interval_secs: 10,
+            ..Default::default()
+        });
+        node_builder
+            .set_esplora_server(esplora_server_url.to_string())
+            .set_gossip_source_p2p();
+        let Some(data_dir_str) = data_dir.to_str() else {
+            return Err(anyhow::anyhow!("Invalid data dir path"));
+        };
+        node_builder.set_storage_dir_path(data_dir_str.to_string());
+
+        let node = Arc::new(node_builder.build()?);
+        node.start().map_err(|e| {
+            error!("Failed to start LDK Node: {e:?}");
+            LightningRpcError::FailedToConnect
+        })?;
+
+        let (htlc_stream_sender, htlc_stream_receiver) = tokio::sync::mpsc::channel(1024);
+
+        let node_clone = node.clone();
+        let htlc_stream_seeder_task_handle = spawn(
+            "ldk lightning node incoming htlc stream seeder",
+            async move {
+                loop {
+                    if let Err(e) =
+                        Self::seed_route_htlcs_stream(&node_clone, &htlc_stream_sender).await
```

Comment:
`seed_route_htlcs_stream` only fails in one spot and all you do is log....you can push that log statement into the function and get rid of the error handling here

Reply:
Excellent point! Done.